### PR TITLE
ORI-482: converge wn handling

### PR DESCRIPTION
**Pull request by nsirola**
Decode the Nt and N4 fields from GLO ephemeris (pulling in librtcm https://github.com/swift-nav/librtcm/pull/78) and compute the GLO TOE from those. 

Add a small unit test with a small RTCM log with ephemerides from all constellations (captured from `RTCM3EPH` stream)

### CODE REVIEW

File: c/tests/check_rtcm3.c [hunk 1]
```diff
@@ -436,6 +436,62 @@ void sbp_callback_eph_wn_rollover(
...
+      toc.wn = msg->toc.wn;
+      toc.tow = msg->toc.tow;
+      break;
+    }
+    case SBP_MSG_EPHEMERIS_GLO: {
+      msg_ephemeris_glo_t *msg = (msg_ephemeris_glo_t *)buffer;
+      if (msg->common.sid.code == CODE_GLO_L1OF &&
+          (msg->common.sid.sat == 8 || msg->common.sid.sat == 22)) {
+        /* these two look otherwise valid but have n4=3 which is 12 years off */
+        return;
```
Comment (nsirola) [inline c/tests/check_rtcm3.c]: this is weird, could not determine if the problem is in the `RTCM3EPH` stream or the decoder

File: c/src/rtcm3_sbp_ephemeris.c [hunk 2]
```diff
@@ -265,11 +265,18 @@ void rtcm3_gps_eph_to_sbp(rtcm_msg_eph *msg_eph,
...
+  glo_time_t glo_time = {0};
+  u32 tb_s = msg_eph->glo.t_b * SEC_IN_15MINUTES;
+  glo_time.nt = msg_eph->glo.nt;
+  glo_time.n4 = msg_eph->glo.n4;
+  glo_time.h = tb_s / HOUR_SECS;
+  glo_time.m = (tb_s - glo_time.h * HOUR_SECS) / MINUTE_SECS;
+  glo_time.s = tb_s - (glo_time.h * HOUR_SECS) - (glo_time.m * MINUTE_SECS);
+
+  /* TODO: construct utc_params struct from state->leap_seconds */
+  gps_time_t toe = glo2gps(&glo_time, NULL);
```
Comment (nsirola) [inline c/src/rtcm3_sbp_ephemeris.c]: proper `utc_params` here would also need the timestamp of the next/previous leap second event that is not accessible here

File: c/src/rtcm3_sbp_ephemeris.c [hunk 3]
```diff
@@ -400,6 +407,6 @@ void rtcm3_bds_eph_to_sbp(rtcm_msg_eph *msg_eph,
   toc.tow =
       (msg_eph->kepler.toc * BEIDOU_TOC_RESOLUTION) + BDS_SECOND_TO_GPS_SECOND;
   normalize_gps_time(&toc);
-  sbp_bds_eph->toc.wn = toc.wn;
+  sbp_bds_eph->toc.wn = gps_adjust_week_cycle(toc.wn, GPS_WEEK_REFERENCE);
```
Comment (mbavaro) [inline c/src/rtcm3_sbp_ephemeris.c]: ouch, thanks for fixing this

File: c/src/rtcm3_sbp_ephemeris.c (hunk 2 above)
Comment (mbavaro) [inline c/src/rtcm3_sbp_ephemeris.c]: nice use of `glo2gps()` :+1:

File: c/tests/check_rtcm3.c (hunk 1 above)
Comment (nsirola) [reply c/tests/check_rtcm3.c]: Looks like the RTCM3EPH has had some problems with the N4 field before: https://software.rtcm-ntrip.org/ticket/103 (and based on this discussion I'm not entirely sure it got fixed then)

Comment (nsirola) [reply c/tests/check_rtcm3.c]: Also some receivers do not fill the additional data fields (including N4) at all. I wonder if the RTCM3EPH problem could result in some sort of majority voting when combining the ephemerides across receivers..

**Comment by nsirola** [comment]
This approach does not work universally as not all RTCM streams provide the N4 or even NT fields for constructing the GLO timestamp properly. Reverting to deducing the year and DOY from current GPS time stamp as seems to be the standard in the business. Thus closing this PR in favor of https://github.com/swift-nav/gnss-converters/pull/191

**Comment by mbavaro** [comment]
I think it's worth double checking with RTCM3EPH and get back to MGEX on that.
Let me give it a shot.

**Comment by mbavaro** [comment]
Used this https://github.com/swift-nav/gnss-converters/pull/149/commits/6f1451aa4b91681128ff85948cd4e509a078584b
to output eph readings:

[eph_r.txt](https://github.com/swift-nav/gnss-converters/files/3113889/eph_r.txt)

**Comment by nsirola** [comment]
I don't have IGS credentials so I don't know if their stream is different, but both the EUREF and SNIP streams show the same satellites giving the wrong N4 value, so they most likely come from the same source:
```
$ curl -s -N -H "Ntrip-Version: Ntrip/2.0" -A "NTRIP curl/7.47.0" <redacted>:<redacted>@www.euref-ip.net:2101/RTCM3EPH | rtcm3tosbp > /dev/null
R01 N4: 6, NT: 1211, t_b: 51, health: 000
R02 N4: 3, NT: 1211, t_b: 51, health: 000
R03 N4: 3, NT: 1211, t_b: 51, health: 000
R04 N4: 6, NT: 1211, t_b: 51, health: 000
R05 N4: 6, NT: 1211, t_b: 51, health: 010
R06 N4: 6, NT: 1211, t_b: 51, health: 000
R07 N4: 6, NT: 1211, t_b: 51, health: 000
R08 N4: 6, NT: 1211, t_b: 51, health: 000
R09 N4: 6, NT: 1211, t_b: 51, health: 000
R10 N4: 6, NT: 1211, t_b: 51, health: 000
R11 N4: 6, NT: 1211, t_b: 51, health: 000
R12 N4: 6, NT: 1211, t_b: 51, health: 000
R13 N4: 3, NT: 1211, t_b: 51, health: 000
R14 N4: 6, NT: 1211, t_b: 51, health: 010
R15 N4: 3, NT: 1211, t_b: 51, health: 000
R16 N4: 6, NT: 1211, t_b: 51, health: 000
R17 N4: 6, NT: 1211, t_b: 51, health: 000
R18 N4: 6, NT: 1211, t_b: 51, health: 000
R19 N4: 6, NT: 1211, t_b: 51, health: 000
R20 N4: 6, NT: 1211, t_b: 51, health: 000
R21 N4: 6, NT: 1211, t_b: 51, health: 000
R22 N4: 6, NT: 1211, t_b: 51, health: 000
R23 N4: 3, NT: 1211, t_b: 51, health: 000
R24 N4: 3, NT: 1211, t_b: 51, health: 000
R26 N4: 6, NT: 1211, t_b: 51, health: 100
^C
$ curl -s -N -H "Ntrip-Version: Ntrip/2.0" -A "NTRIP curl/7.47.0" ntrip.use-snip.com:2101/RTCM3EPH | src/rtcm3tosbp > /dev/null
R26 N4: 6, NT: 1211, t_b: 51, health: 100
R01 N4: 6, NT: 1211, t_b: 51, health: 000
R02 N4: 3, NT: 1211, t_b: 51, health: 000
R03 N4: 3, NT: 1211, t_b: 51, health: 000
R04 N4: 6, NT: 1211, t_b: 51, health: 000
R05 N4: 6, NT: 1211, t_b: 51, health: 010
R06 N4: 6, NT: 1211, t_b: 51, health: 000
R07 N4: 6, NT: 1211, t_b: 51, health: 000
R08 N4: 6, NT: 1211, t_b: 51, health: 000
R09 N4: 6, NT: 1211, t_b: 51, health: 000
R10 N4: 6, NT: 1211, t_b: 51, health: 000
R11 N4: 6, NT: 1211, t_b: 51, health: 000
R12 N4: 6, NT: 1211, t_b: 51, health: 000
R13 N4: 3, NT: 1211, t_b: 51, health: 000
R14 N4: 6, NT: 1211, t_b: 51, health: 010
R15 N4: 3, NT: 1211, t_b: 51, health: 000
R16 N4: 6, NT: 1211, t_b: 51, health: 000
R17 N4: 6, NT: 1211, t_b: 51, health: 000
R18 N4: 6, NT: 1211, t_b: 51, health: 000
R19 N4: 6, NT: 1211, t_b: 51, health: 000
R20 N4: 6, NT: 1211, t_b: 51, health: 000
R21 N4: 6, NT: 1211, t_b: 51, health: 000
R22 N4: 6, NT: 1211, t_b: 51, health: 000
R23 N4: 3, NT: 1211, t_b: 51, health: 000
R24 N4: 3, NT: 1211, t_b: 51, health: 000
```
(printed out with https://github.com/swift-nav/librtcm/blob/nsirola/glo-eph-test/c/src/eph_decode.c#L195)

**Pull request closed without merging**